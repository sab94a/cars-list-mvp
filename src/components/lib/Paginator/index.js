// @flow

import React from 'react';
import qs from 'query-string';
import cx from 'classnames';
import Link from 'components/lib/Link';
import type { Paginator as Props } from 'components/lib/types';
import styles from './index.module.scss';

class Paginator extends React.PureComponent<Props> {
    static Query = 'page'
    static Titles = {
        First: 'First',
        Prev: 'Previous',
        Next: 'Next',
        Last: 'Last'
    }

    buildQuerySearch(page: number):?string {
        const { total, active } = this.props;

        if (page === active || page < 1 || page > total) {
            return null;
        }

        let params = qs.parse(window.location.search);

        return `?${qs.stringify({
            ...params,
            [Paginator.Query]: page
        })}`
    }

    renderLink(link: ?string, title:string) {
        return link ? (
            <Link className={ styles.link } href={ link }>{ title }</Link>
        ) : (
            <span className={ cx(styles.link, styles.disable) }>{ title }</span>
        )
    }

    render() {
        const { active, total } = this.props;

        return (
            <div className={ styles.root }>
                { this.renderLink(this.buildQuerySearch(1), Paginator.Titles.First) }
                { this.renderLink(this.buildQuerySearch(active - 1), Paginator.Titles.Prev) }
                <div className={ styles.content }>{`Page ${ active } of ${ total }`}</div>
                { this.renderLink(this.buildQuerySearch(active + 1), Paginator.Titles.Next) }
                { this.renderLink(this.buildQuerySearch(total), Paginator.Titles.Last) }
            </div>
        )
    }
}

export default Paginator;
