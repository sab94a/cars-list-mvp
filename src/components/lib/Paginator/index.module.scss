@import 'components/lib/variables.scss';

.root {
    display: flex;
    justify-content: center;
    padding: $offset-md 0;
}

.content {
    padding: $offset-md;
}

.link {
    padding: $offset-md;
    text-decoration: none;
    color: $primary;
    cursor: pointer;

    &.disable {
        color: $dark;

        &:active, &:focus {
            color: $dark;
        }
    }

    &:active, &:focus {
        color: $primary-pressed;
    }
}
