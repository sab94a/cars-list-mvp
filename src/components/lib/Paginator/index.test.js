import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Paginator from './';


describe('<Paginator />', () => {
    let component;

    const { location } = window;

    const create = params => {
        component = mount(<Paginator { ...params } />);
    };

    const originalLocation = window.location;

    beforeEach(() => {
        delete window.location;

        window.location = {
            search: '',
        };
    });

    afterEach(() => {
        window.location = originalLocation;
    });

    it('Should render correctly', () => {
        create({ active: 2, total: 10 });
        
        const $content = component.find('.content');
        const $first = component.find('.root').childAt(0);
        const $prev = component.find('.root').childAt(1);
        const $next = component.find('.root').childAt(3);
        const $last = component.find('.root').childAt(4);

        expect($content.text()).toBe('Page 2 of 10');
        expect($first.text()).toBe(Paginator.Titles.First);
        expect($prev.text()).toBe(Paginator.Titles.Prev);
        expect($next.text()).toBe(Paginator.Titles.Next);
        expect($last.text()).toBe(Paginator.Titles.Last);


        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should have correct links', () => {
        create({ active: 3, total: 10 });

        const $first = component.find('.root').childAt(0);
        const $prev = component.find('.root').childAt(1);
        const $next = component.find('.root').childAt(3);
        const $last = component.find('.root').childAt(4);

        expect($first.prop('href')).toBe('?page=1');
        expect($prev.prop('href')).toBe('?page=2');
        expect($next.prop('href')).toBe('?page=4');
        expect($last.prop('href')).toBe('?page=10');
    });

    it('If location have other search, it should add page to current seach', () => {
        window.location.search = '?color=width';

        create({ active: 3, total: 10 });

        const $first = component.find('.root').childAt(0);

        expect($first.prop('href')).toBe('?color=width&page=1');

        window.location.search = '';
    })

    it('Should render SPAN instead of Link if active page 1', () => {
        create({ active: 1, total: 10 });

        const $first = component.find('.root').childAt(0);
        const $prev = component.find('.root').childAt(1);

        expect($first.type()).toBe('span');
        expect($prev.type()).toBe('span');
    });

    it('Should render SPAN instead of Link if active page is last', () => {
        create({ active: 1, total: 10 });

        const $next = component.find('.root').childAt(3);
        const $last = component.find('.root').childAt(4);

        expect($next.type()).toBe('span');
        expect($last.type()).toBe('span');
    });
});