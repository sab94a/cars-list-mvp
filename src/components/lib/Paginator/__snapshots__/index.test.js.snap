// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> Should render correctly 1`] = `
<Paginator
  active={2}
  total={10}
>
  <div
    className="root"
  >
    <Link
      className="link"
      href="?page=1"
    >
      <a
        className="link"
        href="?page=1"
      >
        First
      </a>
    </Link>
    <Link
      className="link"
      href="?page=1"
    >
      <a
        className="link"
        href="?page=1"
      >
        Previous
      </a>
    </Link>
    <div
      className="content"
    >
      Page 2 of 10
    </div>
    <Link
      className="link"
      href="?page=3"
    >
      <a
        className="link"
        href="?page=3"
      >
        Next
      </a>
    </Link>
    <Link
      className="link"
      href="?page=10"
    >
      <a
        className="link"
        href="?page=10"
      >
        Last
      </a>
    </Link>
  </div>
</Paginator>
`;
